toolchain/check/handle_impl.cpp

```diff
@@ -99,7 +107,36 @@ static auto ExtendImpl(Context& context, Parse::AnyImplDeclId parse_node,
     context.emitter().Emit(parse_node, ExtendImplOutsideClass);
     return;
   }
-  auto& enclosing_scope = context.name_scopes().Get(enclosing_scope_id);
+
+  if (params_node.is_valid()) {
+    CARBON_DIAGNOSTIC(ExtendImplForall, Error,
+                      "Cannot `extend` a parameterized `impl`.");
+    context.emitter().Emit(extend_node, ExtendImplForall);
+    enclosing_scope.has_error = true;
+    return;
+  }
+
+  if (context.parse_tree().node_kind(self_type_node) ==
+      Parse::NodeKind::TypeImplAs) {
+    CARBON_DIAGNOSTIC(ExtendImplSelfAs, Error,
+                      "Cannot `extend` an `impl` with an explicit self type.");
+    CARBON_DIAGNOSTIC(ExtendImplSelfAsDefault, Note,
+                      "Remove the explicit `Self` type here.");
+    auto diag = context.emitter().Build(extend_node, ExtendImplSelfAs);
+    if (self_type_id == GetDefaultSelfType(context)) {
+      // If the explicit self type is the default, suggest removing it and
+      // recover.
+      if (auto self_as = context.parse_tree().ExtractAs<Parse::TypeImplAs>(
+              self_type_node)) {
+        diag.Note(self_as->type_expr, ExtendImplSelfAsDefault);
+      }
+      diag.Emit();
+    } else {
+      diag.Emit();
+      enclosing_scope.has_error = true;
+      return;
```

Comment:
This return seems a little odd, why didn't you structure this as:

```
if (self_type_id == ...) {
  ... (dropping emit)
} else {
  enclosing_scope.has_error = true;
}
diag.Emit()
```

Reply:
The first arm of this `if` recovers from the error and does not return.